agent/hcp/telemetry/otel_sink.go

```diff
@@ -0,0 +1,189 @@
+package telemetry
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/agent/hcp/client"
+
+	gometrics "github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	"go.opentelemetry.io/otel/attribute"
+	otelmetric "go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/instrument"
+	otelsdk "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+type OTELSinkOpts struct {
+	Reader otelsdk.Reader
+	Logger hclog.Logger
+}
+
+type OTELSink struct {
+	spaceReplacer *strings.Replacer
+	logger        hclog.Logger
+
+	meterProvider *otelsdk.MeterProvider
+	meter         *otelmetric.Meter
+
+	gaugeStore *gaugeStore
+
+	gaugeInstruments     map[string]*instrument.Float64ObservableGauge
+	counterInstruments   map[string]*instrument.Float64Counter
+	histogramInstruments map[string]*instrument.Float64Histogram
+
+	mutex sync.Mutex
+}
+
+func NewOTELReader(client client.MetricsClient, endpoint string, exportInterval time.Duration) otelsdk.Reader {
+	exporter := NewOTELExporter(client, endpoint)
+	return otelsdk.NewPeriodicReader(exporter, otelsdk.WithInterval(exportInterval))
+}
+
+func NewOTELSink(opts *OTELSinkOpts) (*OTELSink, error) {
+	if opts.Logger == nil {
+		return nil, fmt.Errorf("failed to init OTEL sink: provide valid OTELSinkOpts Logger")
+	}
+
+	if opts.Reader == nil {
+		return nil, fmt.Errorf("failed to init OTEL sink: provide valid OTELSinkOpts Reader")
+	}
+
+	// Setup OTEL Metrics SDK to aggregate, convert and export metrics periodically.
+	res := resource.NewSchemaless()
+	meterProvider := otelsdk.NewMeterProvider(otelsdk.WithResource(res), otelsdk.WithReader(opts.Reader))
+	meter := meterProvider.Meter("github.com/hashicorp/consul/agent/hcp/telemetry")
+
+	gs := &gaugeStore{
+		store: make(map[string]*gaugeValue, 0),
+		mutex: sync.Mutex{},
+	}
+
+	return &OTELSink{
+		spaceReplacer:        strings.NewReplacer(" ", "_"),
+		logger:               opts.Logger.Named("otel_sink"),
+		meterProvider:        meterProvider,
+		meter:                &meter,
+		mutex:                sync.Mutex{},
```

Comment:
You don't need to initialise a mutex in Go.
```suggestion
```

Reply:
Thanks - good point ! fixed this